src/rust-macro-section/numeric-retrievals.md

```diff
@@ -0,0 +1,85 @@
+# Getting or combining subslices of bytes
+
+Some of the macros defined in the Win32 SDK are used to take smaller slices of bytes from larger ones. For example `LOWORD` and `HIWORD` to get the most and least significant `WORD`s from 32-bit integer types serving as packed data. Helpers like these are luckily rather trivial to implement and only require some bitmasking.
+
+## Getting window `LPARAM` values
+
+```rs
+/// Get the low order word as a signed 32-bit integer.
+#[inline(always)]
+pub fn get_x_lparam(lp: windows::Win32::Foundation::LPARAM) -> i32 {
+    lp.0 as i32
+}
+
+/// Get the high order word as a signed 32-bit integer.
+#[inline(always)]
+pub fn get_y_lparam(lp: windows::Win32::Foundation::LPARAM) -> i32 {
+    // the sign bit is duplicated for right shifts on signed ints
+    (lp.0 >> 16) as i32
+}
```

Comment:
Going by [the docs on MSDN](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types#long_ptr) this should be an `__int64` on 64-bit arch. Will add a `cfg` gate there if needed. I can't quickly find anything about the upper 32 bits being zeroed out though, would you happen to have something to confirm that?

Reply:
`WPARAM` and `LPARAM` are type-erased. Their underlying types are meaningless when interpreting the stored value. For a `WM_MOUSEMOVE` message, say, only the low 32 bits of data carry meaningful information, with the remaining bits (if any) unspecified. A quick test uncovered that the system indeed sign-extends the `LPARAM` for negative Y coordinates here. Still, I will attribute that to an implementation detail rather than a contractual guarantee.

For what it's worth, the longer version I proposed [compiles](https://godbolt.org/z/65aMT6WPz) to smaller code. Whether that translates to faster code or slower code, I cannot predict. It's certainly based on fewer hidden assumptions, so that is valuable, even if nothing else.

Reply:
Looking at the output I wonder if we should cast before shifting. The operations are either SHR or SAR; SHR will erase the sign bit and set it to 0, whereas SAR will keep the sign bit from the original full-width int; which would always be 0 if the most significant bits are unused for `LPARAM`s. As the macros call `LOWORD` and `HIWORD` for X and Y respectively, I took a look at those.
And they cast to a `DWORD_PTR` (u32) before doing anything; that'd be an interesting angle to take and it matches up with your assumption of behavior being correct only on 32-bit